File: csharp/Services/AppScanner.cs
using System.Diagnostics;
using System.Runtime.InteropServices;
using System.Text;
using Microsoft.UI.Xaml.Media.Imaging;
using Windows.ApplicationModel;
using Windows.ApplicationModel.Core;
using Windows.Foundation;
using Windows.Graphics.Imaging;
using Windows.Management.Deployment;
using WinCast.Models;

namespace WinCast.Services;

internal static class AppScanner
{
    [ComImport, Guid("5b0d3235-4dba-4d44-865e-8f1d0e4fd04d")]
    [InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
    private interface IMemoryBufferByteAccess
    {
        unsafe int GetBuffer(out byte* value, out uint capacity);
    }

    internal static List<AppItem> ScanApps()
    {
        var apps = new List<AppItem>();

        // Scan Win32 Start Menu shortcuts
        string systemStartMenu = Environment.GetFolderPath(Environment.SpecialFolder.CommonStartMenu) + "\\Programs";
        string userStartMenu = Environment.GetFolderPath(Environment.SpecialFolder.StartMenu) + "\\Programs";

        ScanStartMenuFolder(systemStartMenu, apps);
        ScanStartMenuFolder(userStartMenu, apps);

        // Scan UWP packages
        try
        {
            var packageManager = new PackageManager();
            var packages = packageManager.FindPackagesForUser("");
            foreach (var package in packages)
            {
                if (package.IsFramework) continue;
                try
                {
                    if (OperatingSystem.IsWindowsVersionAtLeast(10, 0, 19041))
                    {
                        foreach (var entry in package.GetAppListEntries())
                        {
                            string name = entry.DisplayInfo?.DisplayName ?? string.Empty;
                            string aumid = entry.AppUserModelId ?? string.Empty;
                            if (string.IsNullOrEmpty(name) || string.IsNullOrEmpty(aumid)) continue;

                            SoftwareBitmap? icon = GetUwpAppIcon(entry);

                            // Dedup: skip if same AUMID exists, or same name with non-UWP already present
                            if (!apps.Any(a => a.AUMID == aumid || (a.Name == name && !a.IsUWP)))
                                apps.Add(new AppItem { Name = name, AUMID = aumid, Path = aumid, IsUWP = true, Icon = icon });
                        }
                    }
                }
                catch { /* Ignore per-package errors */ }
            }
        }
        catch { /* PackageManager might fail */ }

        apps.Sort((a, b) => string.Compare(a.Name, b.Name, StringComparison.OrdinalIgnoreCase));
        return apps;
    }

    private static void ScanStartMenuFolder(string folderPath, List<AppItem> apps)
    {
        if (!Directory.Exists(folderPath)) return;

        try
        {
            foreach (string lnkPath in Directory.EnumerateFiles(folderPath, "*.lnk", new EnumerationOptions
            {
                RecurseSubdirectories = true,
                IgnoreInaccessible = true
            }))
            {
                if (ParseShortcut(lnkPath, out AppItem appItem))
                {
                    string lowerName = appItem.Name.ToLowerInvariant();
                    if (!lowerName.Contains("uninstall") && !lowerName.Contains("read me") &&
                        !lowerName.Contains("readme") && !lowerName.Contains("documentation") &&
                        !lowerName.Contains("help") && !lowerName.Contains("manual"))
                    {
                        if (!apps.Any(a => a.Path == appItem.Path && a.Name == appItem.Name))
                            apps.Add(appItem);
                    }
                }
            }
        }
        catch { /* iteration errors */ }
    }

    private static bool ParseShortcut(string lnkPath, out AppItem appItem)
    {
        appItem = new AppItem();
        NativeMethods.IShellLinkW? shellLink = null;
        try
        {
            shellLink = (NativeMethods.IShellLinkW)new NativeMethods.ShellLinkWrapper();
            var persistFile = (NativeMethods.IPersistFile)shellLink;
            persistFile.Load(lnkPath, NativeMethods.STGM_READ);

            appItem.Name = Path.GetFileNameWithoutExtension(lnkPath);

            var sb = new StringBuilder(260);
            shellLink.GetPath(sb, 260, IntPtr.Zero, 0);
            appItem.Path = sb.ToString();

            // Try to get AUMID from property store
            NativeMethods.PropVariant propVar = default;
            bool propVarNeedsClear = false;
            try
            {
                if (shellLink is NativeMethods.IPropertyStore propStore)
                {
                    var pkey = new NativeMethods.PropertyKey
                    {
                        fmtid = NativeMethods.FMTID_PKEY_AppUserModel_ID,
                        pid = NativeMethods.PID_AppUserModel_ID
                    };
                    propStore.GetValue(ref pkey, out propVar);
                    propVarNeedsClear = true;
                    if (propVar.vt == NativeMethods.VT_LPWSTR && propVar.pwszVal != IntPtr.Zero)
                    {
                        appItem.AUMID = Marshal.PtrToStringUni(propVar.pwszVal) ?? string.Empty;
                        appItem.IsUWP = true;
                    }
                }
            }
            catch { /* Property store might not be available */ }
            finally
            {
                if (propVarNeedsClear)
                {
                    NativeMethods.PropVariantClear(ref propVar);
                }
            }

            // Extract icon
            appItem.Icon = ExtractShortcutIcon(shellLink, lnkPath, appItem.Path);

            return !string.IsNullOrEmpty(appItem.Path) || appItem.IsUWP;
        }
        catch
        {
            return false;
        }
        finally
        {
            if (shellLink != null)
            {
                Marshal.ReleaseComObject(shellLink);
            }
        }
    }

    private static SoftwareBitmap? ExtractShortcutIcon(NativeMethods.IShellLinkW shellLink, string lnkPath, string targetPath)
    {
        IntPtr hIcon = IntPtr.Zero;
        try
        {
            // Try shortcut icon location first
            var sb = new StringBuilder(260);
            shellLink.GetIconLocation(sb, 260, out int iconIndex);
            if (sb.Length > 0)
            {
                string expandedPath = Environment.ExpandEnvironmentVariables(sb.ToString());
                NativeMethods.ExtractIconExW(expandedPath, iconIndex, out hIcon, out _, 1);
            }

            // Fallback: extract from target executable
            if (hIcon == IntPtr.Zero && !string.IsNullOrEmpty(targetPath))
                NativeMethods.ExtractIconExW(targetPath, 0, out hIcon, out _, 1);

            // Final fallback: shell file info
            if (hIcon == IntPtr.Zero)
            {
                var sfi = new NativeMethods.SHFILEINFOW();
                NativeMethods.SHGetFileInfoW(lnkPath, 0, ref sfi, (uint)Marshal.SizeOf(sfi),
                    NativeMethods.SHGFI_ICON | NativeMethods.SHGFI_LARGEICON);
                hIcon = sfi.hIcon;
            }

            if (hIcon != IntPtr.Zero)
                return GetBitmapFromHIcon(hIcon);

            return null;
        }
        finally
        {
            if (hIcon != IntPtr.Zero)
                NativeMethods.DestroyIcon(hIcon);
        }
    }

    private static SoftwareBitmap? GetBitmapFromHIcon(IntPtr hIcon)
    {
        if (hIcon == IntPtr.Zero) return null;

        if (!NativeMethods.GetIconInfo(hIcon, out var iconInfo)) return null;
        try
        {
            NativeMethods.GetObject(iconInfo.hbmColor, Marshal.SizeOf<NativeMethods.BITMAP>(), out var bm);
            int width = bm.bmWidth;
            int height = bm.bmHeight;
            if (width <= 0 || height <= 0) return null;

            var bmi = new NativeMethods.BITMAPINFO();
            bmi.bmiHeader.biSize = (uint)Marshal.SizeOf<NativeMethods.BITMAPINFOHEADER>();
            bmi.bmiHeader.biWidth = width;
            bmi.bmiHeader.biHeight = -height; // top-down
            bmi.bmiHeader.biPlanes = 1;
            bmi.bmiHeader.biBitCount = 32;
            bmi.bmiHeader.biCompression = NativeMethods.BI_RGB;

            byte[] pixelData = new byte[width * height * 4];

            IntPtr hdc = NativeMethods.GetDC(IntPtr.Zero);
            try
            {
                NativeMethods.GetDIBits(hdc, iconInfo.hbmColor, 0, (uint)height, pixelData, ref bmi, NativeMethods.DIB_RGB_COLORS);
            }
            finally
            {
                NativeMethods.ReleaseDC(IntPtr.Zero, hdc);
            }

            // Fix alpha channel: if all alpha bytes are 0, set them to 255
            int pixelCount = width * height;
            bool hasAlpha = false;
            for (int i = 3; i < pixelCount * 4; i += 4)
            {
                if (pixelData[i] != 0) { hasAlpha = true; break; }
            }
            if (!hasAlpha)
            {
                for (int i = 3; i < pixelCount * 4; i += 4)
                    pixelData[i] = 255;
            }

            // Convert byte array to SoftwareBitmap via stream
            using var stream = new Windows.Storage.Streams.InMemoryRandomAccessStream();
            var encoder = BitmapEncoder.CreateAsync(BitmapEncoder.PngEncoderId, stream).GetAwaiter().GetResult();
            encoder.SetPixelData(BitmapPixelFormat.Bgra8, BitmapAlphaMode.Premultiplied,
                (uint)width, (uint)height, 96.0, 96.0, pixelData);
            encoder.FlushAsync().GetAwaiter().GetResult();

            stream.Seek(0);
            var decoder = BitmapDecoder.CreateAsync(stream).GetAwaiter().GetResult();
            return decoder.GetSoftwareBitmapAsync(BitmapPixelFormat.Bgra8, BitmapAlphaMode.Premultiplied).GetAwaiter().GetResult();
        }
        finally
        {
            if (iconInfo.hbmColor != IntPtr.Zero)
                NativeMethods.DeleteObject(iconInfo.hbmColor);
            if (iconInfo.hbmMask != IntPtr.Zero)
                NativeMethods.DeleteObject(iconInfo.hbmMask);
        }
    }

    private static SoftwareBitmap? GetUwpAppIcon(AppListEntry entry)
    {
        try
        {
            var logoStreamRef = entry.DisplayInfo?.GetLogo(new Size(44, 44));
            if (logoStreamRef == null) return null;

            using var stream = logoStreamRef.OpenReadAsync().GetAwaiter().GetResult();
            var decoder = BitmapDecoder.CreateAsync(stream).GetAwaiter().GetResult();
            var softwareBitmap = decoder.GetSoftwareBitmapAsync().GetAwaiter().GetResult();

            if (softwareBitmap.BitmapPixelFormat != BitmapPixelFormat.Bgra8 ||
                softwareBitmap.BitmapAlphaMode != BitmapAlphaMode.Premultiplied)
            {
                softwareBitmap = SoftwareBitmap.Convert(softwareBitmap, BitmapPixelFormat.Bgra8, BitmapAlphaMode.Premultiplied);
            }
            return softwareBitmap;
        }
        catch { return null; }
    }

    internal static bool Launch(AppItem item)
    {
        if (item.IsUWP)
        {
            NativeMethods.IApplicationActivationManager? manager = null;
            try
            {
                var clsidType = Type.GetTypeFromCLSID(NativeMethods.CLSID_ApplicationActivationManager);
                if (clsidType != null)
                {
                    manager = Activator.CreateInstance(clsidType) as NativeMethods.IApplicationActivationManager;
                    if (manager != null)
                    {
                        manager.ActivateApplication(item.AUMID, null, 0, out _);
                        return true;
                    }
                }
                throw new InvalidOperationException("Failed to activate application using ActivationManager.");
            }
            catch
            {
                try
                {
                    Process.Start(new ProcessStartInfo("shell:AppsFolder\\" + item.AUMID) { UseShellExecute = true });
                    return true;
                }
                catch { return false; }
            }
            finally
            {
                if (manager != null)
                {
                    Marshal.ReleaseComObject(manager);
                }
            }
        }
        else
        {
            try
            {
                Process.Start(new ProcessStartInfo(item.Path) { UseShellExecute = true });
                return true;
            }
            catch { return false; }
        }
    }
}

File: AppScanner.cpp
#include "pch.h"
#include "AppScanner.h"
#include <shobjidl.h>
#include <MemoryBuffer.h>

// COM Guid for IMemoryBufferByteAccess
struct __declspec(uuid("5b0d3235-4dba-4d44-865e-8f1d0e4fd04d")) IMemoryBufferByteAccess : ::IUnknown
{
    virtual HRESULT __stdcall GetBuffer(BYTE** value, UINT32* capacity) = 0;
};

// IApplicationActivationManager is already defined in <shobjidl.h>

// Property Key for AUMID in shortcuts
const PROPERTYKEY PKEY_AppUserModel_ID = { { 0x9F4C6855, 0x3779, 0x4F77, { 0x75, 0x1A, 0x88, 0xEC, 0x58, 0x58, 0x99, 0x18 } }, 5 };

namespace wincast::scanner
{
    std::vector<AppItem> AppScanner::ScanApps()
    {
        // Initialize COM on this background thread
        winrt::init_apartment(winrt::apartment_type::multi_threaded);

        std::vector<AppItem> apps;

        // 1. Scan Win32 Start Menu Shortcuts
        // System Start Menu
        WCHAR szPath[MAX_PATH];
        if (SUCCEEDED(SHGetFolderPathW(NULL, CSIDL_COMMON_STARTMENU, NULL, 0, szPath)))
        {
            std::wstring systemStartMenu = std::wstring(szPath) + L"\\Programs";
            ScanStartMenuFolder(systemStartMenu, apps);
        }
        
        // User Start Menu
        if (SUCCEEDED(SHGetFolderPathW(NULL, CSIDL_STARTMENU, NULL, 0, szPath)))
        {
            std::wstring userStartMenu = std::wstring(szPath) + L"\\Programs";
            ScanStartMenuFolder(userStartMenu, apps);
        }

        // 2. Scan UWP Packaged Apps
        try
        {
            winrt::Windows::Management::Deployment::PackageManager packageManager;
            auto packages = packageManager.FindPackagesForUser(L"");
            for (auto const& package : packages)
            {
                if (!package.IsFramework())
                {
                    try
                    {
                        auto appEntries = package.GetAppListEntries();
                        for (auto const& entry : appEntries)
                        {
                            AppItem item;
                            item.Name = entry.DisplayInfo().DisplayName();
                            item.AUMID = entry.AppUserModelId();
                            item.IsUWP = true;
                            
                            // Get logo stream in background
                            item.Icon = GetUwpAppIcon(entry);
                            
                            // Filter out empty name apps
                            if (!item.Name.empty() && !item.AUMID.empty())
                            {
                                // Check if we already have this app
                                auto it = std::find_if(apps.begin(), apps.end(), [&](AppItem const& existing) {
                                    return existing.AUMID == item.AUMID || (existing.Name == item.Name && !existing.IsUWP);
                                });
                                if (it == apps.end())
                                {
                                    apps.push_back(item);
                                }
                            }
                        }
                    }
                    catch (...)
                    {
                        // Ignore package errors
                    }
                }
            }
        }
        catch (...)
        {
            // PackageManager might fail in some configurations
        }

        // Sort applications alphabetically
        std::sort(apps.begin(), apps.end(), [](AppItem const& a, AppItem const& b) {
            return a.Name < b.Name;
        });

        return apps;
    }

    void AppScanner::ScanStartMenuFolder(std::wstring const& folderPath, std::vector<AppItem>& apps)
    {
        if (!std::filesystem::exists(folderPath)) return;

        try
        {
            for (auto const& dirEntry : std::filesystem::recursive_directory_iterator(folderPath, std::filesystem::directory_options::skip_permission_denied))
            {
                if (dirEntry.is_regular_file())
                {
                    std::wstring path = dirEntry.path().wstring();
                    if (path.size() > 4 && path.substr(path.size() - 4) == L".lnk")
                    {
                        AppItem appItem;
                        if (ParseShortcut(path, appItem))
                        {
                            // Filter out uninstallers, manuals, readme etc.
                            std::wstring lowerName = appItem.Name;
                            std::transform(lowerName.begin(), lowerName.end(), lowerName.begin(), ::towlower);
                            
                            if (lowerName.find(L"uninstall") == std::wstring::npos &&
                                lowerName.find(L"read me") == std::wstring::npos &&
                                lowerName.find(L"readme") == std::wstring::npos &&
                                lowerName.find(L"documentation") == std::wstring::npos &&
                                lowerName.find(L"help") == std::wstring::npos &&
                                lowerName.find(L"manual") == std::wstring::npos)
                            {
                                // Avoid duplicates
                                auto it = std::find_if(apps.begin(), apps.end(), [&](AppItem const& existing) {
                                    return existing.Path == appItem.Path && existing.Name == appItem.Name;
                                });
                                if (it == apps.end())
                                {
                                    apps.push_back(appItem);
                                }
                            }
                        }
                    }
                }
            }
        }
        catch (...)
        {
            // Path iteration errors
        }
    }

    bool AppScanner::ParseShortcut(std::wstring const& lnkPath, AppItem& appItem)
    {
        wil::com_ptr<IShellLinkW> psl;
        if (FAILED(CoCreateInstance(CLSID_ShellLink, NULL, CLSCTX_INPROC_SERVER, IID_IShellLinkW, psl.put_void())))
        {
            return false;
        }

        wil::com_ptr<IPersistFile> ppf;
        if (FAILED(psl->QueryInterface(IID_IPersistFile, ppf.put_void())))
        {
            return false;
        }

        if (FAILED(ppf->Load(lnkPath.c_str(), STGM_READ)))
        {
            return false;
        }

        // Get Name from the shortcut file name
        std::filesystem::path fsPath(lnkPath);
        appItem.Name = fsPath.stem().wstring();

        // Get Path
        WCHAR szTarget[MAX_PATH];
        psl->GetPath(szTarget, MAX_PATH, NULL, 0);
        appItem.Path = szTarget;

        // Extract description
        WCHAR szDesc[MAX_PATH];
        if (SUCCEEDED(psl->GetDescription(szDesc, MAX_PATH)))
        {
            // Can be used if needed
        }

        // Check if UWP AUMID is stored in the shortcut (e.g. for modern apps)
        wil::com_ptr<IPropertyStore> pPropStore;
        if (SUCCEEDED(psl->QueryInterface(IID_IPropertyStore, pPropStore.put_void())))
        {
            PROPVARIANT propvar;
            PropVariantInit(&propvar);
            if (SUCCEEDED(pPropStore->GetValue(PKEY_AppUserModel_ID, &propvar)))
            {
                if (propvar.vt == VT_LPWSTR && propvar.pwszVal != nullptr)
                {
                    appItem.AUMID = propvar.pwszVal;
                    appItem.IsUWP = true;
                }
                PropVariantClear(&propvar);
            }
        }

        // Extract Icon
        HICON hIcon = nullptr;
        // Try getting icon from the shortcut itself
        WCHAR szIconPath[MAX_PATH];
        int iconIndex = 0;
        if (SUCCEEDED(psl->GetIconLocation(szIconPath, MAX_PATH, &iconIndex)) && szIconPath[0] != L'\0')
        {
            // Expand environment variables in path
            WCHAR szExpandedPath[MAX_PATH];
            ExpandEnvironmentStringsW(szIconPath, szExpandedPath, MAX_PATH);
            ExtractIconExW(szExpandedPath, iconIndex, &hIcon, nullptr, 1);
        }

        // Fallback: Extract icon from the target executable
        if (!hIcon && !appItem.Path.empty())
        {
            ExtractIconExW(appItem.Path.c_str(), 0, &hIcon, nullptr, 1);
        }

        // Final Fallback: Shell File Info (handles folders, documents, etc.)
        if (!hIcon)
        {
            SHFILEINFOW sfi = {};
            if (SHGetFileInfoW(lnkPath.c_str(), 0, &sfi, sizeof(sfi), SHGFI_ICON | SHGFI_LARGEICON))
            {
                hIcon = sfi.hIcon;
            }
        }

        if (hIcon)
        {
            appItem.Icon = GetBitmapFromHIcon(hIcon);
            DestroyIcon(hIcon);
        }

        // Only return true if we have a path to run, or if it's UWP
        return !appItem.Path.empty() || appItem.IsUWP;
    }

    winrt::Windows::Graphics::Imaging::SoftwareBitmap AppScanner::GetBitmapFromHIcon(HICON hIcon)
    {
        if (!hIcon) return nullptr;

        ICONINFO iconInfo;
        if (!GetIconInfo(hIcon, &iconInfo)) return nullptr;

        wil::unique_hbitmap hbmColor(iconInfo.hbmColor);
        wil::unique_hbitmap hbmMask(iconInfo.hbmMask);

        BITMAP bm;
        GetObject(hbmColor.get(), sizeof(bm), &bm);

        int width = bm.bmWidth;
        int height = bm.bmHeight;

        try
        {
            winrt::Windows::Graphics::Imaging::SoftwareBitmap softwareBitmap(
                winrt::Windows::Graphics::Imaging::BitmapPixelFormat::Bgra8,
                width,
                height,
                winrt::Windows::Graphics::Imaging::BitmapAlphaMode::Premultiplied);

            auto buffer = softwareBitmap.LockBuffer(winrt::Windows::Graphics::Imaging::BitmapBufferAccessMode::Write);
            auto reference = buffer.CreateReference();
            auto byteAccess = reference.as<IMemoryBufferByteAccess>();

            byte* data = nullptr;
            uint32_t capacity = 0;
            if (SUCCEEDED(byteAccess->GetBuffer(&data, &capacity)))
            {
                BITMAPINFO bmi = {};
                bmi.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
                bmi.bmiHeader.biWidth = width;
                bmi.bmiHeader.biHeight = -height; // Top-down
                bmi.bmiHeader.biPlanes = 1;
                bmi.bmiHeader.biBitCount = 32;
                bmi.bmiHeader.biCompression = BI_RGB;

                HDC hdc = GetDC(nullptr);
                GetDIBits(hdc, hbmColor.get(), 0, height, data, &bmi, DIB_RGB_COLORS);
                ReleaseDC(nullptr, hdc);
                
                // Fix alpha channel if it's completely zero (some icons lack alpha information and will render invisible)
                bool hasAlpha = false;
                for (int i = 3; i < width * height * 4; i += 4)
                {
                    if (data[i] != 0)
                    {
                        hasAlpha = true;
                        break;
                    }
                }

                if (!hasAlpha)
                {
                    for (int i = 3; i < width * height * 4; i += 4)
                    {
                        data[i] = 255;
                    }
                }
            }

            return softwareBitmap;
        }
        catch (...)
        {
            return nullptr;
        }
    }

    winrt::Windows::Graphics::Imaging::SoftwareBitmap AppScanner::GetUwpAppIcon(winrt::Windows::ApplicationModel::Core::AppListEntry const& entry)
    {
        try
        {
            auto logoStreamRef = entry.DisplayInfo().GetLogo(winrt::Windows::Foundation::Size{ 44, 44 });
            if (!logoStreamRef) return nullptr;

            auto stream = logoStreamRef.OpenReadAsync().get();
            auto decoder = winrt::Windows::Graphics::Imaging::BitmapDecoder::CreateAsync(stream).get();
            auto softwareBitmap = decoder.GetSoftwareBitmapAsync().get();

            if (softwareBitmap.BitmapPixelFormat() != winrt::Windows::Graphics::Imaging::BitmapPixelFormat::Bgra8 ||
                softwareBitmap.BitmapAlphaMode() != winrt::Windows::Graphics::Imaging::BitmapAlphaMode::Premultiplied)
            {
                softwareBitmap = winrt::Windows::Graphics::Imaging::SoftwareBitmap::Convert(
                    softwareBitmap,
                    winrt::Windows::Graphics::Imaging::BitmapPixelFormat::Bgra8,
                    winrt::Windows::Graphics::Imaging::BitmapAlphaMode::Premultiplied);
            }
            return softwareBitmap;
        }
        catch (...)
        {
            return nullptr;
        }
    }

    bool AppScanner::Launch(AppItem const& item)
    {
        if (item.IsUWP)
        {
            wil::com_ptr<IApplicationActivationManager> pActivationManager;
            HRESULT hr = CoCreateInstance(
                CLSID_ApplicationActivationManager,
                NULL,
                CLSCTX_INPROC_SERVER,
                IID_IApplicationActivationManager,
                pActivationManager.put_void());

            if (SUCCEEDED(hr))
            {
                DWORD processId = 0;
                hr = pActivationManager->ActivateApplication(item.AUMID.c_str(), NULL, static_cast<ACTIVATEOPTIONS>(0), &processId);
                return SUCCEEDED(hr);
            }
            
            // Fallback via protocol
            try
            {
                std::wstring protocol = L"shell:AppsFolder\\" + item.AUMID;
                ShellExecuteW(NULL, L"open", protocol.c_str(), NULL, NULL, SW_SHOWNORMAL);
                return true;
            }
            catch (...)
            {
                return false;
            }
        }
        else
        {
            HINSTANCE hInst = ShellExecuteW(NULL, L"open", item.Path.c_str(), NULL, NULL, SW_SHOWNORMAL);
            return (reinterpret_cast<INT_PTR>(hInst) > 32);
        }
    }
}
